src/Interactive/Features/Interactive/Core/InteractiveHost.cs

```diff
@@ -226,27 +222,38 @@ private bool CheckAlive(Process process)
 
         ~InteractiveHost()
         {
-            Dispose(joinThreads: false, disposing: false);
+            Dispose(disposing: false);
         }
 
         public void Dispose()
         {
-            Dispose(joinThreads: false, disposing: true);
+            Dispose(disposing: true);
         }
 
-        private void Dispose(bool joinThreads, bool disposing)
+        internal void Dispose(bool disposing)
         {
+            var semaphore = _disposeSemaphore;
+            if (semaphore != null && disposing)
+            {
+                semaphore.Wait();
+                semaphore.Dispose();
+            }
```

Comment:
Let's move this right after the ```_lazyRemoteService.Dispose(disposing)``` call since the remote service is the one signaling the semaphore.